homeassistant/components/motionmount/config_flow.py

```diff
@@ -0,0 +1,175 @@
+            info[CONF_UUID] = config_entries.DEFAULT_DISCOVERY_UNIQUE_ID
+
+        # If the device mac is valid we use it, otherwise we use the default id
+        if info.get(CONF_UUID, EMPTY_MAC) != EMPTY_MAC:
+            unique_id = info[CONF_UUID]
+        else:
+            unique_id = config_entries.DEFAULT_DISCOVERY_UNIQUE_ID
+
+        name = info.get(CONF_NAME, user_input[CONF_HOST])
+
+        await self.async_set_unique_id(unique_id)
+        self._abort_if_unique_id_configured(
+            updates={
+                CONF_HOST: user_input[CONF_HOST],
+                CONF_PORT: user_input[CONF_PORT],
+            }
+        )
+
+        return self.async_create_entry(title=name, data=user_input)
+
+    async def async_step_zeroconf(
+        self, discovery_info: zeroconf.ZeroconfServiceInfo
+    ) -> FlowResult:
+        """Handle zeroconf discovery."""
+
+        # Extract information from discovery
+        host = discovery_info.hostname
+        port = discovery_info.port
+        zctype = discovery_info.type
+        name = discovery_info.name.removesuffix(f".{zctype}")
+        unique_id = discovery_info.properties.get("mac")
+
+        self.discovery_info.update(
+            {
+                CONF_HOST: host,
+                CONF_PORT: port,
+                CONF_NAME: name,
+            }
+        )
+
+        if unique_id:
+            # If we already have the unique id, try to set it now
+            # so we can avoid probing the device if its already
+            # configured or ignored
+            await self.async_set_unique_id(unique_id)
+            self._abort_if_unique_id_configured(
+                updates={CONF_HOST: host, CONF_PORT: port}
+            )
+        else:
+            # Avoid probing devices that already have an entry
+            self._async_abort_entries_match({CONF_HOST: host})
+
+        self.context.update({"title_placeholders": {"name": name}})
+
+        try:
+            info = await self._validate_input(self.discovery_info)
+        except (ConnectionError, socket.gaierror):
+            return self.async_abort(reason="cannot_connect")
+        except TimeoutError:
+            return self.async_abort(reason="time_out")
+        except motionmount.NotConnectedError:
+            return self.async_abort(reason="not_connected")
+        except motionmount.MotionMountResponseError:
+            info = {}
+            # We continue as we want to be able to connect with older FW that does not support MAC address
+
+        # If the device supplied as with a valid MAC we use that
+        if info.get(CONF_UUID, EMPTY_MAC) != EMPTY_MAC:
+            unique_id = info[CONF_UUID]
+
+        if unique_id:
+            await self.async_set_unique_id(unique_id)
+            self._abort_if_unique_id_configured(
+                updates={CONF_HOST: host, CONF_PORT: port}
+            )
+
+        await self._async_handle_discovery_without_unique_id()
```

Comment:
It looks like this needs a branch here to not call `_async_handle_discovery_without_unique_id` if we have a unique id?

Reply:
I'm not sure.
The IPP integration (on which I heavily relied during development) follows the same construct.
The call to `_async_handle_discovery_without_unique_id()` performs a check whether `self.unique_id` is set and returns if so.
I think it's a matter of taste to branch here or have `_async_handle_discovery_without_unique_id()` perform the check. I'm fine with both ways, which do you prefer?

Reply:
I'd generally prefer to not call a function if we know its going to return and do nothing. Any code we can avoid executing is less potential for bugs